### PR TITLE
Add document and test case for SM2 cert verification

### PR DESCRIPTION
**Pull request by InfoHunter**
This follows the #8321, to add related document and test cases for 2 newly added APIs and 2 new options in apps/verify.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by InfoHunter** [comment]
CI failure is relevant, fixing...

File: apps/verify.c [hunk 1]
```diff
@@ -174,6 +175,7 @@ int verify_main(int argc, char **argv)
             break;
         case OPT_SM2HEXID:
             /* try to parse the input as hex string first */
+            sm2_free = 1;
```
Comment (mlindner) [inline apps/verify.c]: Instead of using a flag, instead set sm2_id to NULL after this expression and then you can simply unconditionally free.

Comment (InfoHunter) [reply apps/verify.c]: after which expression?

Comment (mlindner) [reply apps/verify.c]: On line 183, following the `}` of the `if (sm2_id == NULL) {`

**Comment by mlindner** [comment]
Somewhere add a note that X509_get0_sm2_id "returns an internal pointer that shouldn't be freed". This is done for all the `get0` functions, from my understanding.

**Comment by mlindner** [comment]
I left a comment on https://github.com/openssl/openssl/pull/8321 that there's apparently a memory corruption issue if the user uses the `X509_set_sm2_id` function.

File: apps/verify.c (hunk 1 above)
Comment (InfoHunter) [reply apps/verify.c]: I don't quite get the point, I think the conditionally free action at the `end` section is protecting the `sm2_id` set by `OPT_SM2ID` from being freed, which should not be freed in any case.

**Comment by InfoHunter** [comment]
Err, I need to check what other similar APIs do

File: apps/verify.c (hunk 1 above)
Comment (mlindner) [reply apps/verify.c]: Sorry disregard my comment, I was misreading the code.

Comment (InfoHunter) [reply apps/verify.c]: Okay

**Comment by InfoHunter** [comment]
> Somewhere add a note that X509_get0_sm2_id "returns an internal pointer that shouldn't be freed". This is done for all the get0 functions, from my understanding.

After a while of thinking I found the `get0` in the API name indicates the return value should not be freed by the caller, so it seems not necessary to put a line in the doc to describe that behaviour.

**Comment by InfoHunter** [comment]
> I left a comment on #8321 that there's apparently a memory corruption issue if the user uses the X509_set_sm2_id function.

Hmmm, should this be a problem that API provider need to handle?

**Comment by mattcaswell** [comment]
> After a while of thinking I found the get0 in the API name indicates the return value should not be freed by the caller, so it seems not necessary to put a line in the doc to describe that behaviour.

That is an OpenSSL naming peculiarity that isn't obvious to someone not familiar to OpenSSL. We should always document the memory semantics in these get0/get1 functions IMO.

**Comment by mlindner** [comment]
> > I left a comment on #8321 that there's apparently a memory corruption issue if the user uses the X509_set_sm2_id function.
> 
> Hmmm, should this be a problem that API provider need to handle?

Yes, because `X509_set_sm2_id` becomes impossible to use with heap allocated structures. At the very least if you're making such a strange restriction then it MUST be documented that the `X509_set_sm2_id` function ONLY works with stack allocated `ASN1_OCTET_STRING` structures.

**Comment by InfoHunter** [comment]
Leave a statement on the return value of get0 function

**Comment by mlindner** [comment]
Not sure how this works, but is it possible to add a test case for the X509_get0_sm2_id and X509_set_sm2_id functions? If you add a test case for X509_set_sm2_id with heap allocated sm2_id you should see double free.

Example function:
```
void test_func(X509 *x, ASN1_OCTET_STRING *sm2_id) {
    X509_set_sm2_id(x, sm2_id);
    X509_free(x);
    ASN1_OCTET_STRING_free(sm2_id); <--- double free of sm2_id->data
}
```

**Comment by InfoHunter** [comment]
I think current apps/verify.c has already done a case of heap-allocated `sm2-id` if you specify the `sm2-hex-id` option on the command line - and which is included in the updated test case in this PR.

**Comment by InfoHunter** [comment]
Ahh, sorry that I missed your point, did you mean if the following code is changed to heap-allocated:

```
    if (sm2id != NULL) {
#ifndef OPENSSL_NO_SM2
        ASN1_OCTET_STRING v;
        v.data = sm2id;
        v.length = sm2idlen;
        X509_set_sm2_id(x, &v);
#endif
    }
```
?

**Comment by InfoHunter** [comment]
Besides, the `sm2_id` field in x509 structure is not handled in the x509 ASN1 callback, so theoretically it should not be freed twice.

**Comment by InfoHunter** [comment]
Besides, the `sm2_id` field in x509 structure is not handled in the x509 ASN1 callback, so theoretically it should not be freed twice.

**Comment by mlindner** [comment]
> Besides, the `sm2_id` field in x509 structure is not handled in the x509 ASN1 callback, so theoretically it should not be freed twice.

Then you have an equivalent problem.

```
void test_func(X509 *x, ASN1_OCTET_STRING *sm2_id) {
    X509_set_sm2_id(x, sm2_id);
    ASN1_OCTET_STRING_free(sm2_id); <--- free of internal x->sm2_id->data buffer
}
```

If you say that you should not free sm2_id ever, then sm2_id will leak.

File: doc/man1/verify.pod [hunk 2]
```diff
@@ -316,6 +318,16 @@ Display information about the certificate chain that has been built (if
 successful). Certificates in the chain that came from the untrusted list will be
 flagged as "untrusted".
 
+=item B<-sm2-id>
+
+Specifies the ID string when verifying an SM2 certificate. The ID string is
```
Comment (paulidale) [inline doc/man1/verify.pod]: First sentence beginning with:

> Specify the ID string to use when verifying

File: doc/man1/verify.pod [hunk 3]
```diff
@@ -316,6 +318,16 @@ Display information about the certificate chain that has been built (if
...
+required by the SM2 signature algorithm for signing and verification.
+
+=item B<-sm2-hex-id>
+
+Specifies the SM2 ID string in binary form. The only valid input of this option
```
Comment (paulidale) [inline doc/man1/verify.pod]: Replace this paragraph with:

> Specify a binary ID string to use when signing or verifying using an SM2 certificate. The argument for this option is string of hexadecimal digits.

File: doc/man3/X509_get0_sm2_id.pod [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+=head1 DESCRIPTION
+
+X509_get0_sm2_id() gets the ID value of an SM2 certificate B<x> by returning an
+B<ASN1_OCTET_STRING> object which should not be freed by the caller.
+X509_set_sm2_id() sets the B<sm2_id> value to an SM2 certificate B<x>.
+
+=head1 NOTES
+
+SM2 signature algorithm requires an ID value when generating and verifying a
+signature. Functions described in this manual provides the users the ability
```
Comment (paulidale) [inline doc/man3/X509_get0_sm2_id.pod]: Second sentence becomes:

> The functions described in this manual provide the user with the ability to set and retrieve the SM2 ID value.

**Comment by InfoHunter** [comment]
Wording is addressed in the latest commit

**Comment by InfoHunter** [comment]
> Then you have an equivalent problem.

In the example you provided, the caller should free the *sm2_id*, instead of inside the `test_func`

**Comment by InfoHunter** [comment]
Merged to master, thanks!

**Pull request closed without merging**